### PR TITLE
Notify the claimant of their application

### PR DESCRIPTION
**Pull request by robbiepaul**
So that a user has confidence that they have finished a thing, and so they have a record of having completed their application, we will send the user an email confirming their claim.

We want to include a reference number so that it is easy to find their claim if they need support.

- [delayed_job](https://github.com/collectiveidea/delayed_job) to process background jobs
- [GOV.UK Notify](https://www.notifications.service.gov.uk/) for sending ActionMailer emails

The content used is just temporary and is from the confirmation page.

## Preview of the email

Screenshot 2019-06-13 at 15 13 44

### CODE REVIEW

**Comment by robbiepaul** [comment]
[Notify the claimant of their application](https://trello.com/c/0VdIxBKg/185-notify-the-claimant-of-their-application)

**Comment by robbiepaul** [comment]
I need to add some instructions to the readme. If you need an invite to Notify let me know

File: Gemfile.lock [hunk 1]
```diff
@@ -113,6 +114,9 @@ GEM
       nokogiri (>= 1.5.9)
     mail (2.7.1)
       mini_mime (>= 0.1.1)
+    mail-notify (0.1.0)
```
Comment (erbridge) [inline Gemfile.lock]: Looks like this is an old version of `mail-notify`. I wonder why your Bundler is pulling in old dependencies.

File: app/controllers/claims_controller.rb [hunk 2]
```diff
@@ -106,6 +106,10 @@ def claim_params
     )
   end
 
+  def claim_submitted
```
Comment (erbridge) [inline app/controllers/claims_controller.rb]: Maybe this could be named after what it does rather than when it happens? e.g. `mail_claim_submitted` or something.

File: spec/requests/claims_spec.rb [hunk 3]
```diff
@@ -178,6 +181,10 @@
             expect(in_progress_claim.submitted_at).to be_present
           end
 
+          it "sends an email" do
+            expect(claim_mailer_spy).to have_received(:deliver_later)
+          end
+
```
Comment (erbridge) [inline spec/requests/claims_spec.rb]: Think we could also test the negative case: we don't send an email if submission fails.

File: Gemfile.lock (hunk 1 above)
Comment (robbiepaul) [reply Gemfile.lock]: Hmm, strange. I'll have a look

Comment (robbiepaul) [reply Gemfile.lock]: It looks like this is the most recent version: https://rubygems.org/gems/mail-notify

Comment (erbridge) [reply Gemfile.lock]: Oh. Hah. Sorry. Read it as `0.0.1`.

**Comment by robbiepaul** [comment]
@erbridge I've made those changes, ready for another look

File: config/application.rb [hunk 4]
```diff
@@ -32,5 +32,8 @@ class Application < Rails::Application
     config.generators do |g|
       g.orm :active_record, primary_key_type: :uuid
     end
+
+    config.active_job.queue_adapter = :que
+    config.action_mailer.deliver_later_queue_name = ""
```
Comment (mec) [inline config/application.rb]: I read that this should be `"default"` is that implied somehow?

Comment (robbiepaul) [reply config/application.rb]: Yeah, you're right. `:default` works as well, I've changed it so it's more obvious.

**Comment by robbiepaul** [comment]
@mec I must've missed that first time round reading the docs, it isn't as obvious on the pre 1.0 branch. 1.0 as still in beta: https://github.com/chanks/que/tree/0.x

It says it's highly recommended that we switch. We'll have to make sure we're all on the same version of Postgres or we'll get issues like we saw before. Maybe Que isn't the right solution?

**Comment by mec** [comment]
@robbiepaul we should all talk, like we did last time. It seems like we will always use PG so it's a given, but the version problems might be an issue, I'd want to know more before we made any jump?

**Comment by mec** [comment]
It does say 'You can use schema_format as :ruby if you want but keep in mind this is highly advised against, as some parts of Que will not work.' – worth understanding which 'parts' first?

File: db/schema.rb [hunk 5]
```diff
@@ -28,6 +28,17 @@
     t.index ["code"], name: "index_local_authority_districts_on_code", unique: true
   end
 
+  create_table "que_jobs", primary_key: ["queue", "priority", "run_at", "job_id"], comment: "3", force: :cascade do |t|
```
Comment (mec) [inline db/schema.rb]: None of this schema looks too scary?

Comment (robbiepaul) [reply db/schema.rb]: Yeah I agree, I'll switch it to `sql` locally just to see what we'd miss by keeping as `ruby`

Comment (mec) [reply db/schema.rb]: If we stick with `ruby` be good to note this in the commit message for future us?

Comment (robbiepaul) [reply db/schema.rb]: This is what the change that is produced by Que looks like when switching to `sql`:

```
CREATE TABLE public.que_jobs (
    priority smallint DEFAULT 100 NOT NULL,
    run_at timestamp with time zone DEFAULT now() NOT NULL,
    job_id bigint NOT NULL,
    job_class text NOT NULL,
    args json DEFAULT '[]'::json NOT NULL,
    error_count integer DEFAULT 0 NOT NULL,
    last_error text,
    queue text DEFAULT ''::text NOT NULL
);

COMMENT ON TABLE public.que_jobs IS '3';

CREATE SEQUENCE public.que_jobs_job_id_seq
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;

ALTER SEQUENCE public.que_jobs_job_id_seq OWNED BY public.que_jobs.job_id;
```
The only difference I see is the `CREATE SEQUENCE public.que_jobs_job_id_seq...` bit

Comment (robbiepaul) [reply db/schema.rb]: Yeah, looks like the `job_id` is always 0 with `:ruby` https://github.com/chanks/que/issues/148

Comment (robbiepaul) [reply db/schema.rb]: Not sure what to do about this, I'll start a thread on Slack

Comment (robbiepaul) [reply db/schema.rb]: Opted for [`delayed_job`](https://github.com/collectiveidea/delayed_job) instead

File: spec/requests/claims_spec.rb [hunk 6]
```diff
@@ -165,6 +165,12 @@
       end
 
       context "when updating from check-your-answers" do
+        let(:claim_mailer_spy) { spy(:mailer) }
+        before do
+          allow(ClaimMailer).to receive(:submitted).with(in_progress_claim)
```
Comment (tekin) [inline spec/requests/claims_spec.rb]: Rather than stubbing the mailer I think it's better to check the contents of `ActionMailer::Base.deliveries` as part of the expectation: tests the expected outcome, not the implementation.

**Comment by robbiepaul** [comment]
@tekin I've changed the request spec to check the contents of `ActionMailer::Base.deliveries`

I've set up a new API key for staging that will only send to 'Team and whitelist' and I've put the environment variables in to Heroku
